torch/csrc/api/include/torch/nn/cursor.h

```diff
@@ -0,0 +1,157 @@
+  // on whether `T` is const.
+  using Iterator = typename std::conditional<
+      std::is_const<T>::value,
+      typename std::vector<Item>::const_iterator,
+      typename std::vector<Item>::iterator>::type;
+
+  // No need for a virtual destructor, as cursors are not intended to be used
+  // polymorhpically (i.e. we are relying on non-virtual inheritance).
+
+  // Note that these functions may only be called on lvalues (that's the
+  // ampersand next to the function)! This prevents code like `auto iterator =
+  // module.modules().begin()`, since `iterator` would be pointing to a `vector`
+  // that gets destructed at the end of the expression. This is not a problem
+  // for range loops, as they capture the range expression (the thing to the
+  // right of the colon in `for (auto x : ...)`) before iteration.
+  Iterator begin() & noexcept;
+  Iterator end() & noexcept;
+
+  /// Applies a function to every *value* available. The function should accept
+  /// a single argument, that is a reference to the value type (e.g. `Module&`).
+  template <typename Function>
+  void apply(const Function& function) {
+    for (auto module : *this) {
+      function(*module);
+    }
+  }
+
+  /// Applies a function to every *item* available. The function should accept
+  /// two arguments, one taking a reference to the key type (always `const
+  /// std::string&`) and the other taking a reference to the value type (e.g.
+  /// `Module&`).
+  template <typename Function>
+  void apply_items(const Function& function) {
+    for (auto module : *this) {
+      function(module.key, *module);
+    }
+  }
+
+  /// Applies a function to every *value* available, and stores the return value
+  /// of the function into the iterator. The function should accept
+  /// a single argument, that is a reference to the value type (e.g. `Module&`).
+  template <typename Iterator, typename Function>
+  void map(Iterator output_iterator, Function function) {
+    for (auto module : *this) {
+      *output_iterator = function(*module);
+    }
+  }
+
+  /// Applies a function to every *value* available, and stores the return value
+  /// of the function into the iterator. The function should accept
+  /// two arguments, one taking a reference to the key type (always `const
+  /// std::string&`) and the other taking a referen
+  template <typename Iterator, typename Function>
+  void map_items(Iterator output_iterator, Function function) {
+    for (auto module : *this) {
+      *output_iterator = function(module.key, *module);
+    }
+  }
+
+  /// Attempts to find a value for the given `key`. If found, returns a pointer
+  /// to the value. If not, returns a null pointer.
+  T* find(const std::string& key) noexcept;
+
+  /// Returns true if an item with the given `key` exists.
+  bool contains(const std::string& key) noexcept;
+
+  /// Counts the number of items available.
+  size_t size() const noexcept;
+
+ protected:
+  std::vector<Item> items_;
```

Comment:
Why do `Cursor`s keep vectors of items? I thought they would just traverse the tree

Reply:
So I had this first, but it adds some complication. Primarily, doing iterative DFS requires parent pointers, and that makes it super hard/dangerous to do things like moves (since all the parent pointers have to be updated). I discussed this with Zach, and we'll never really have more than, say, 100 elements in the hierarchy right (and maybe 10 in practice)? So highly likely this will be fast enough for anyone (and the code is so much simpler). So I figured we'll start with the simple solution, and can always add back the more complex one if it's unbearable

Reply:
Right, let's keep it.